Reintroduce query plan reuse during warmup (#5255)
apollographql/router: .changesets/fix_geal_fix_hashing_algorithm.md

### Fix cache key hashing algorithm ([Issue #5160](https://github.com/apollographql/router/issues/5160))

> [!IMPORTANT]
> If you have enabled [Distributed query plan caching](https://www.apollographql.com/docs/router/configuration/distributed-caching/#distributed-query-plan-caching), this release changes the hashing algorithm used for the cache keys.  On account of this, you should anticipate additional cache regeneration cost when updating between these versions while the new hashing algorithm comes into service.

> The Router contains a schema aware query hashing algorithm, designed to return the same hash across schema updates if they do not affect the query. It was used mainly to avoid planning a lot of queries when warming up the query plan cache. It was deactivated due to a regression. This reintroduces this algorithm, with a serie of reliability and performance fixes, along with better observability.

Reviewer: ```suggestion
The Router contains a schema aware query hashing algorithm, designed to return the same hash across schema updates if they do not affect the query. It was used mainly to avoid planning a lot of queries when warming up the query plan cache. It was deactivated due to a regression. This reintroduces this algorithm, with a number of reliability and performance fixes, along with better observability.
```

> The hashing algorithm can be activated in configuration, if query plan cache warm up is already enabled:

Reviewer: I don't understand why cache warm up is mentioned here. There's no connection is there?

```suggestion
The hashing algorithm can be activated in configuration.
```

Author: it needs the warmup to be usable

> ```yaml title="router.yaml"
> supergraph:
>   query_planning:
>     warmed_up_queries: 100
>     experimental_reuse_query_plans: true
> ```

Reviewer: If removing the reference to warm-up, also remove the configuration.

```suggestion
```yaml title="router.yaml"
supergraph:
  query_planning:
    experimental_reuse_query_plans: true
```
```

Author: same as above, it is not usable without the warmup

There is a counter metric named `apollo.router.query.planning.warmup.reused` that can be used to track the hashing algorithm benefits:
- if the `experimental_reuse_query_plans` option is false, the `query_plan_reuse_active` metric attribute will be false. Cache warm up will not reuse query plans according to the algorithm, but it will evaluate if some of them could have been reused and report that in the metric

> - if the `experimental_reuse_query_plans` option is true, then the `query_plan_reuse_active` metric attribute will be true

Reviewer: Since this isn't the main thrust of the changelog (which is about the fixed key hashing algorithm), we should introduce this. Something like:

```
We've also introduced changed to our metrics that are helpful in tracking the impact of these changes.
```

> Fixes included in this change:
> - strenghten the hashing algorithm to prevent collisions. In particular, the query string is always hashed in, making sure that different queries cannot get the same hash

Reviewer: ```suggestion
Summary:
- strengthen the hashing algorithm to prevent collisions. In particular, the query string is always hashed in, making sure that different queries cannot get the same hash
```
